File: src/two-pointer/pair-sum.ts
// O(n)
export function pairSum(nums: number[], target: number): number[] {
    let start = 0,
        end = nums.length - 1

    while (start < end) {
        const currentSum = nums[start] + nums[end]
        if (currentSum < target) {
            start += 1
        } else if (currentSum > target) {
            end -= 1
        } else {
            return [start, end]
        }
    }

    return []
}

File: src/two-pointer/pair_sum.go
package twopointer

// Time complexity: O(n)
// Space complexity: O(1)
func PairSum(nums []int, target int) []int {
	start := 0
	end := len(nums) - 1

	for start < end {
		currentSum := nums[start] + nums[end]
		if currentSum < target {
			start++
		} else if currentSum > target {
			end--
		} else {
			return []int{start, end}
		}
	}

	return []int{}
}
